src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -104,52 +111,92 @@ class CodeMenuRenderer implements IListRenderer<ICodeActionMenuItem, ICodeAction
 		data.disposables = [];
 		data.root = container;
 		data.text = document.createElement('span');
-		// data.detail = document.createElement('');
+
+		const iconContainer = document.createElement('div');
+		iconContainer.className = 'icon-container';
+
+		data.icon = document.createElement('div');
+
+		iconContainer.append(data.icon);
+		container.append(iconContainer);
 		container.append(data.text);
-		// container.append(data.detail);
 
 		return data;
 	}
 	renderElement(element: ICodeActionMenuItem, index: number, templateData: ICodeActionMenuTemplateData): void {
 		const data: ICodeActionMenuTemplateData = templateData;
-		const text = element.action.label;
+
 		const isSeparator = element.isSeparator;
+		const isHeader = element.isHeader;
 
-		element.isEnabled = element.action.enabled;
+		if (isSeparator) {
+			data.root.classList.add('separator');
+			data.root.style.height = '10px';
+		} else if (isHeader) {
+			const text = element.headerTitle;
+			data.text.textContent = text;
+			element.isEnabled = false;
+			data.root.classList.add('group-header');
+		} else {
+			const text = element.action.label;
+			element.isEnabled = element.action.enabled;
 
-		if (element.action instanceof CodeActionAction) {
+			if (element.action instanceof CodeActionAction) {
 
-			// Check documentation type
-			element.isDocumentation = element.action.action.kind === CodeActionMenu.documentationID;
-			if (!element.isDocumentation) {
+				// Check documentation type
+				element.isDocumentation = element.action.action.kind === CodeActionMenu.documentationID;
 
-				// Check if action has disabled reason
-				if (element.action.action.disabled) {
-					data.root.title = element.action.action.disabled;
+				if (element.isDocumentation) {
+					data.text.textContent = text;
+					data.root.classList.add('documentation');
 				} else {
-					const updateLabel = () => {
-						const [accept, preview] = this.acceptKeybindings;
-						data.root.title = localize({ key: 'label', comment: ['placeholders are keybindings, e.g "F2 to Refactor, Shift+F2 to Preview"'] }, "{0} to Refactor, {1} to Preview", this.keybindingService.lookupKeybinding(accept)?.getLabel(), this.keybindingService.lookupKeybinding(preview)?.getLabel());
-					};
-					updateLabel();
+					// Icons and Label modifaction based on group
+					const group = element.action.action.kind;
+
+					if (CodeActionKind.Refactor.equals(new CodeActionKind(String(group)))) {
+						data.icon.className = Codicon.symbolArray.classNames;
+						const label = text.replace(/(Surround With )*(Surround With: )*/g, '');
+						data.text.textContent = label.charAt(0).toUpperCase() + label.slice(1);
```

Comment:
I don't think we should try transforming the labels like this. This logic won't work if the labels are localized so best to try being dumb. We can add new Apis to better support these going forward